pycbc/psd/variation.py

```diff
@@ -210,10 +210,187 @@ def find_trigger_value(psd_var, idx, start, sample_rate):
+    fweight[0] = 0.
+    norm = (sum(abs(fweight) ** 2) / (len(fweight) - 1.)) ** -0.5
+    fweight *= norm
+    fwhiten = numpy.sqrt(2. / sample_rate) / numpy.sqrt(plong)
+    fwhiten[0] = 0.
+    full_filt = sig.hann(int(psd_duration * sample_rate)) * numpy.roll(
+        irfft(fwhiten * fweight), int(psd_duration / 2) * sample_rate)
+
+    return full_filt
+
+
+def live_calc_psd_variation(strain,
+                            full_filt,
+                            increment,
+                            data_trim=2.0,
+                            short_stride=0.25):
+    """
+    Calculate the psd variation in the PyCBC Live search.
+
+    The Live strain data is convolved with the filter to produce a timeseries
+    containing the PSD variation values for each sample. The mean square of
+    the timeseries is calculated over the short_stride to find outliers caused
+    by short duration glitches. Outliers are replaced with the average of
+    adjacent elements in the array. This array is then further averaged every
+    second to produce the PSD variation timeseries.
+
+    Parameters
+    ----------
+    strain : pycbc.timeseries
+        Live data being searched through by the PyCBC Live search.
+    full_filt : numpy.ndarray
+        A filter created by `live_create_filter`.
+    increment : float
+        The number of seconds in each increment in the PyCBC Live search.
+    data_trim : float
+        The number of seconds to be trimmed from either end of the convolved
+        timeseries to prevent artefacts.
+    short_stride : float
+        The number of seconds to average the PSD variation timeseries over to
+        remove the effects of short duration glitches.
+
+    Returns
+    -------
+    psd_var : pycbc.timeseries
+        A timeseries containing the PSD variation values.
+
+    """
+    sample_rate = int(strain.sample_rate)
+
+    # Grab the last increments worth of data, plus padding for edge effects.
+    astrain = strain.time_slice(strain.end_time - increment - (data_trim * 3),
+                                strain.end_time)
+
+    # Convolve the data and the filter to produce the PSD variation timeseries,
+    #  then trim the beginning and end of the data to prevent edge effects.
+    wstrain = sig.fftconvolve(astrain, full_filt, mode='same')
+    wstrain = wstrain[int(data_trim * sample_rate):-int(data_trim * sample_rate)]
+
+    # Create a PSD variation array by taking the mean square of the PSD
+    #  variation timeseries every short_stride
+    short_ms = numpy.mean(
+        wstrain.reshape(-1, int(sample_rate * short_stride)) ** 2, axis=1)
+
+    # Define an array of averages that is used to substitute outliers
+    #  variation array
+    ave = 0.5 * (short_ms[2:] + short_ms[:-2])
+    outliers = short_ms[1:-1] > (2. * ave)
+    short_ms[1:-1][outliers] = ave[outliers]
+
+    # Calculate the PSD variation every second by averaging over short_ms every
+    #  samples_per_second.
```

Comment:
I'm not sure if 'every samples_per_second' makes sense .. what do you want to say here?

Reply:
Probably doesn't make sense once you convert 'sample_per_second' to a number. 

I want to say that we're averaging the short_ms array every X number of elements to find the average psd variation value each second. samples_per_second depends on the length of short_stride, if short_stride is 0.1 then it is every 10 samples/elements. If short_stride is 0.25 then it's 4 samples_per_second.

Reply:
oh, I think I see. See suggestion ..